nodebuilder/store.go

```diff
@@ -182,3 +182,61 @@ func indexPath(base string) string {
 func dataPath(base string) string {
 	return filepath.Join(base, "data")
 }
+
+// constraintBadgerConfig returns BadgerDB configuration optimized for low memory usage and more frequent
+// compaction which prevents memory spikes.
+// This is particularly important for LNs with restricted memory resources.
+//
+// With the following configuration, a LN uses up to 300iB of RAM during initial sync/sampling
+// and up to 200MiB during normal operation. (on 4 core CPU, 8GiB RAM droplet)
+//
+// With the following configuration and "-tags=jemalloc", a LN uses no more than 180MiB during initial
+// sync/sampling and up to 100MiB during normal operation. (same hardware spec)
+// NOTE: To enable jemalloc, build celestia-node with "-tags=jemalloc" flag, which configures Badger to
+// use jemalloc instead of Go's default allocator.
+//
+// TODO(@Wondertan): Consider alternative less constraint configuration for FN/BN
+// TODO(@Wondertan): Consider dynamic memory allocation based on available RAM
+func constraintBadgerConfig() *dsbadger.Options {
+	opts := dsbadger.DefaultOptions // this must be copied
+	// ValueLog:
+	// 2mib default => 500b - makes sure headers and samples are stored in value log
+	// This *tremendously* reduces the amount of memory used by the node, up to 10 times less during
+	// compaction
+	opts.ValueThreshold = 500
+	// make sure we don't have any limits for stored headers
+	opts.ValueLogMaxEntries = 100000000
+	// run value log GC more often to spread the work over time
+	opts.GcInterval = time.Minute * 1
+	// default 0.5 => 0.125 - makes sure value log GC is more aggressive on reclaiming disk space
+	opts.GcDiscardRatio = 0.125
+	// Snappy saves us ~200MiB of disk space on the mainnet chain to the commit's date
+	// TODO(@Wondertan): Does it worth the overhead?
+	opts.Compression = options.Snappy
+	opts.BlockCacheSize = 16 << 20
+
+	// MemTables:
+	// default 64mib => 16mib - decreases memory usage and makes compaction more often
+	opts.MemTableSize = 16 << 20
+	// default 5 => 3
+	opts.NumMemtables = 3
+	// default 5 => 3
+	opts.NumLevelZeroTables = 3
+	// default 15 => 5 - this prevents memory growth on CPU constraint systems by blocking all writers
+	opts.NumLevelZeroTablesStall = 5
```

Comment:
> NumLevelZeroTablesStall lowered looks risky. When it is reached we are talking about block all Put operations until compaction is done VS using more RAM. What would be the reason to make app stall completely instead of consuming more?

Because consuming more will eventually oom the app

Reply:
> It all comes to the actual usecase: total amount of data stored and speed of new data being written. Lower values for those settings will use less RAM, but could bottleneck on throughput. Would be great to do tests to figure out limitations of suggested config.

The test showed that memory usage is less for each param, as well as NumLevelZeroTables produces mem spikes for compaction more often

Reply:
`NumLevelZeroTables` lowered will trigger compaction earlier, which is totally fine. So it is more of a trade off: more CPU instead of more RAM. We even have it as 1 for inverted index to trigger compaction as fast as possible.

opts.NumLevelZeroTables = 3
opts.NumLevelZeroTablesStall = 5

Still looks dangerously close to me. It is only 2 table apart between starting compaction to being blocked. Would suggest to make them much further apart.

Reply:
I think it's not dangerous for LN. For FN/BN we should have a more relaxed config.

Reply:
> NumLevelZeroTables lowered will trigger compaction earlier, which is totally fine. So it is more of a trade off: more CPU instead of more RAM

as you said, it's all the same CPU time in the end. This only makes sure we spread CPU and RAM load over time